### PR TITLE
Provide cwl_utils.parser.load_* and save to provide unified way to load and save the CWL object for all CWL versions

### PR DESCRIPTION
**Pull request by tom-tan**
This is the first attempt to solve #98 by providing a unified version of `load_*` family.
It contains the following changes:

- Move `parser_v1_*` to the `cwl_utils/parser` directory
- Add a utility function `cwl_version` to `cwl_utils.parser.__init__.py`. It takes YAML object and returns a string for `cwlVersion`. If there are no `cwlVersion` fields in the top level in the YAML object, it returns `None`.
- Add the following functions: `load_document`, `load_document_by_string` and `load_document_by_yaml`. They internally checks the `cwlVersion` field and calls the appropriate load function in `cwl_utils.parser.parser_v1_*`.
  - Note that it sets `cwlVersion` field for each `Process` object when the resulted of `cwl_utils.parser.parser_v1_*` is an array.
- Add the `save` function to `cwl_utils.parser.__init__.py`. It is almost the same as `cwl_utils.parser.parser_v1_*.save`.
  - Note that it returns a packed object with `cwlVersion` and `$graph` fields when the input is an array of `Process` objects. `cwlVersion` field in the resulted dictionary uses the `cwlVersion` field in the first `Process` object in the input array.

This request is just a PoC. Therefore it lacks:
- [x] tests
- [x] formatting
- [x] fixes to other codes such as `cwl_normalizer.py` and `cwl_v1_0_expression_refactor.py`
- [x] fixes to the build systems such as `Makefile`
- [x] docstring

I will be happy if there are suggestions or comments.
If this direction is acceptable, I will add tests and fix other missing changes to make this request ready to merge.

### CODE REVIEW

**Comment by mr-c** [comment]
>     * `cwlVersion` field in the resulted dictionary uses the `cwlVersion` field in the first `Process` object in the input array.

Probably should upgrade all documents to the highest `cwlVersion` present

**Review by mr-c — COMMENTED** [review]
This is great, thanks! Please continue

**Comment by tom-tan** [comment]
> Probably should upgrade all documents to the highest `cwlVersion` present

I guess it should not, according to the current spec.

The [spec of v1.2](https://www.commonwl.org/v1.2/CommandLineTool.html#Packed_documents) says:

> All process objects in a packed document must validate and execute as the `cwlVersion` appearing the top level. A `cwlVersion` field appearing anywhere other than the top level must be ignored.

Therefore, all the documents in the packed object must be validated as the top level `cwlVersion` without upgrading (changing) the documents even if some documents has `cwlVersion: v1.0`.

Fixing the spec is out of scope of this request.

We have to be careful to fix the spec because it may cause a breaking change. That is, it may affects the conformance tests and the implementations that supports the packed objects in the current spec in addition to the document of the spec.

**Comment by mr-c** [comment]
> > Probably should upgrade all documents to the highest `cwlVersion` present
> 
> I guess it should not, according to the current spec.
> 
> The [spec of v1.2](https://www.commonwl.org/v1.2/CommandLineTool.html#Packed_documents) says:
> 
> > All process objects in a packed document must validate and execute as the `cwlVersion` appearing the top level. A `cwlVersion` field appearing anywhere other than the top level must be ignored.

This is not incompatible with my suggestion. Here the spec is talking about the `cwlVersion` of the resulting "packed" document, not the `cwlVersion` of the parts that got packed together.

> 
> Therefore, all the documents in the packed object must be validated as the top level `cwlVersion` without upgrading (changing) the documents even if some documents has `cwlVersion: v1.0`.

Correct. Therefore the safest thing to do is:
1. prior to packing, discover all `cwlVersion`s.
2. Upgrade (if needed) all document to the highest `cwlVersion` using [cwl-upgrader](https://github.com/common-workflow-language/cwl-upgrader/) ([example usage](https://github.com/common-workflow-language/cwl-utils/blob/main/cwl_utils/cwl_normalizer.py#L91))
3. Place this "highest" `cwlVersion` in the root of the new packed document, discarding all others.

This is what `cwltool` does.

> 
> Fixing the spec is out of scope of this request.

I don't think the spec needs fixing in this regard.

**Comment by mr-c** [comment]
> Should `cwl-utils` support non-released versions such as `draft-3`, `draft-4.dev3` and `v1.2.0-dev5`?

Nope! 😃

**Comment by mr-c** [comment]
> > ```
> > * `cwlVersion` field in the resulted dictionary uses the `cwlVersion` field in the first `Process` object in the input array.
> > ```
> 
> Probably should upgrade all documents to the highest `cwlVersion` present

Pardon me. I've been thinking about creating an equivalent of `cwltool --pack` using this codebase and I confused that with your goal: parsing an already `--pack`ed workflow. 

In this instance, you were already correct: no upgrades are required and only the top-level `cwlVersion` value should be used.

Sorry for the confusion!

**Comment by tom-tan** [comment]
I have several points to be discussed:
- I moved `parser_v*` into `cwl_utils/parser`. So now a parser for each cwlVersion is in the module `cwl_utils.parser.parser_v*` but its name seems to be redundant. How about using `cwl_utils.parser.cwl_v*` rather than `cwl_utils.parser.parser_v*`?
- How to do with an invalid inputs (e.g., lists, strings and other non-object values) for `cwl_version`?
  - Return `None`
  - Throw `ValidationException` or other appropriate exceptions
  - Just add a docstring such as `input parameter must be a mapping object`

**Comment by mr-c** [comment]
>     * How about using `cwl_utils.parser.cwl_v*` rather than `cwl_utils.parser.parser_v*`?

Sure!

> How to do with an invalid inputs (e.g., lists, strings and other non-object values) for `cwl_version`?
> 
>     * Return `None`
> 
>     * Throw `ValidationException` or other appropriate exceptions
> 
>     * Just add a docstring such as `input parameter must be a mapping object`

2nd and 3rd option, please

File: tests/test_parser.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+from cwl_utils.parser import cwl_version, load_document, save
```
Comment (mr-c) [inline tests/test_parser.py:1]: This file needs a top level doc string to explain its purpose. You can check it with `make diff_pydocstyle_report`

**Review by mr-c — CHANGES_REQUESTED** [review]
Almost there!

File: tests/test_parser.py (hunk 1 above)
Comment (tom-tan) [reply tests/test_parser.py:2]: Done! Thanks!

**Review by mr-c — APPROVED** [review]

**Comment by mr-c** [comment]
Thank you @tom-tan !

**Comment by tom-tan** [comment]
Thanks!

**Pull request merged**